Proposal for issue #13679: Allow progress tracking of scenario objectives in the park menu

JonasKlose opened the pull request:
- Added progress tracking in Park.cpp
- Moved duplicate code in Scenario.cpp in functions
- Added progress tracking UI strings for en-GB and de-DE
- Allow to print UINT32 numbers in the game UI
- Merge changes with current dev branch

duncanspumpkin commented on data/language/de-DE.txt:
```diff
@@ -3648,6 +3648,15 @@ STR_6396    :Bildschirmschoner und -stromsparen deaktivieren
 STR_6397    :Falls ausgewählt, werden der Bildschirmschoner und andere Bildschirmstromsparfunktionen blockiert, solange OpenRCT2 läuft.
 STR_6398    :Datei enthält nicht unterstützte Fahrgeschäfttypen. Bitte aktualisieren Sie zu einer neueren Version von OpenRCT2.
 STR_6399    :OpenRCT2 benötigt Dateien vom Originalspiel RollerCoaster Tycoon 2, um zu funktionieren. Bitte setzen Sie die „game_path“-Variable in config.ini auf das Verzeichnis, in das Sie RollerCoaster Tycoon 2 installiert haben, dann starten Sie OpenRCT2 neu.
+STR_6400    :{WINDOW_COLOUR_2}Ihr Fortschritt:
```
remove these changes. All non en-GB changes are done on the localisation repo.

duncanspumpkin commented on src/openrct2/localisation/FormatCodes.cpp:
```diff
@@ -35,6 +35,7 @@ static const std::unordered_map<std::string_view, FormatToken> FormatTokenMap =
     { "COMMA2DP32",           FormatToken::Comma2dp32,          },
     { "COMMA16",              FormatToken::Comma16,             },
     { "UINT16",               FormatToken::UInt16,              },
+    { "UINT32",               FormatToken::UInt32,              },
```
Do not add new format codes we are working on removing these.

duncanspumpkin commented on src/openrct2/localisation/FormatCodes.cpp:
```diff
@@ -119,6 +120,7 @@ bool FormatTokenTakesArgument(FormatToken token)
         case FormatToken::Comma2dp32:
         case FormatToken::Comma16:
         case FormatToken::UInt16:
+        case FormatToken::UInt32:
```
remove

duncanspumpkin commented on src/openrct2/localisation/FormatCodes.h:
```diff
@@ -33,6 +33,7 @@ enum class FormatToken
     Comma2dp32,
     Comma16,
     UInt16,
+    UInt32,
```
remove

duncanspumpkin commented on src/openrct2/localisation/Formatting.cpp:
```diff
@@ -483,6 +483,7 @@ namespace OpenRCT2
         switch (token)
         {
             case FormatToken::UInt16:
+            case FormatToken::UInt32:
```
remove

duncanspumpkin commented on test/tests/FormattingTests.cpp:
```diff
@@ -347,6 +347,14 @@ TEST_F(FormattingTests, format_number_basic)
     ASSERT_STREQ("123", ss.data());
 }
 
+TEST_F(FormattingTests, format_number_large)
```
remove

duncanspumpkin commented on src/openrct2/scenario/Scenario.cpp:
```diff
@@ -856,35 +887,15 @@ ObjectiveStatus Objective::CheckMonthlyRideIncome() const
 }
 
-ObjectiveStatus Objective::CheckFinish5RollerCoasters() const
+int Objective::GetFinish5RollerCoastersProgress() const
```
don't use int use int32_t

duncanspumpkin commented:
Remove your formatting changes and use the already available formatting codes.

JonasKlose commented:
> Remove your formatting changes and use the already available formatting codes.

This would imply a downcast of the gNumGuestsInPark variable that is of type **uint32_t**. While I doubt that there will be OpenRCT2 with so many guests, shouldn't the variable converted to **uint16_t** too, then?

duncanspumpkin commented:
You would never output a raw 32bit value. You would always use the comma version. Users would expect the thousands comma and we already have the 32bit comma formatter.

JonasKlose commented:
> You would never output a raw 32bit value. You would always use the comma version. Users would expect the thousands comma and we already have the 32bit comma formatter.

Ah, ok. I'll try to do it that way. The comma formatter looked like it was supposed to only work with 16bit values.

JonasKlose commented:
Ok. I got a bit confused because my compiler didn't want to let go of one of the files.
However, I've successfully reverted all the changes to the Formatter and the connected files, merged in the current dev branch, reverted the changes to the german localization and fixed the code formatting issue as required.

I did test all of the possible changes of the park main window with my personal save games.
All of my save games opened correctly and the park window showed the progress of the scenario.

Sorry for the check fail on the commit message test. I need to learn more about proper commit messages, it seems. :$

JonasKlose commented:
Forgot the required switch from int to int32_t variables.

duncanspumpkin commented:
Okay this is looking much better. Does progress as a unit make more sense than as a percentage? I think it makes sense here but just want to see others opinions.

tupaschoal commented:
> Okay this is looking much better. Does progress as a unit make more sense than as a percentage? I think it makes sense here but just want to see others opinions.

I think it does, you have a direct answer of what you still need to do. As a percentage you might have to do some calculations

JonasKlose commented:
> Okay this is looking much better. Does progress as a unit make more sense than as a percentage? I think it makes sense here but just want to see others opinions.

For the amount of money and guests, progress bars could be a viable alternative.

For smaller progress like for roller coasters, there could be a hint that a certain number of coasters is close to completion. If the length requirement is 1200, this could count the coasters that are between 1000 and 1200, for example.

JonasKlose commented:
https://github.com/OpenRCT2/OpenRCT2/blob/c4c6baff36ccef5e903e981f90ca4d888fde40c8/src/openrct2-ui/windows/Park.cpp#L1393-L1401

I've expanded the window of the Park objectives window to fit the text. However I think it's a bit too large now.
Is there a way to calculate the proper window size, ideally in code? I think the size of the window will depend on the font size, the font itself, the length and width of the texts and the language of the texts.

duncanspumpkin commented:
Size of windows are just educated guesses usually based on en-gb size. If other languages need more space we do a follow up pr to increase the size. Some windows dynamically resize for the text but there isn't much point wasting performance on that for every window.

ocalhoun6 commented:
> there could be a hint that a certain number of coasters is close to completion. If the length requirement is 1200, this could count the coasters that are between 1000 and 1200, for example.

That sounds like it would be getting a bit overly complicated.

IntelOrca commented:
Looks good so far, I think it should have percent or progress bar though. Perhaps something like:
```
Combined income:
    £400 [#####                        ] £1,000
```
using the same progress bars that are drawn on tab 2 of the guest window.

IntelOrca commented on data/language/en-GB.txt:
```diff
@@ -3661,6 +3661,15 @@ STR_6406    :GOG RollerCoaster Tycoon 2 Installer
+STR_6410    :{WINDOW_COLOUR_2}Your Progress:
+STR_6411    :{BLACK}-
+STR_6412    :{BLACK}Guests: {COMMA32} of {COMMA16}
+STR_6413    :{BLACK}Rating: {COMMA16} of {INT32}
```
Different format types used here, should both be `COMMA16`

JonasKlose replied:
The gParkRating variable is a uint16_t, while the target rating are the hard-coded numbers 600 and 700 which should be represented as int32_t. Or should I cast them to uint16_t?

IntelOrca replied:
They should both be comma separated. As for size, I'd make them both 32-bit and cast as necessary.

JonasKlose replied:
OK, I'll do that. :)

IntelOrca commented on data/language/en-GB.txt:
```diff
@@ -3661,6 +3661,15 @@ STR_6406    :GOG RollerCoaster Tycoon 2 Installer
 STR_6409    :The selected file is not the offline GOG Installer for RollerCoaster Tycoon 2. You may have downloaded the GOG Galaxy downloader stub or selected the wrong file.
+STR_6410    :{WINDOW_COLOUR_2}Your Progress:
+STR_6411    :{BLACK}-
+STR_6412    :{BLACK}Guests: {COMMA32} of {COMMA16}
```
Different format types used here, should both be `COMMA32`

JonasKlose replied:
This is correct, according to the code. The gNumGuestsInPark variable is a uint32_t and the Objective NumGuests variable is a uint16_t. I was confused about that, too.

IntelOrca replied:
Okay, the objective num guests will at some point become a `uint32_t`, so maybe you can cast it to int32 for now.

JonasKlose replied:
I've converted to uint32_t for now. I doubt that there'll ever be a negative number of guests and converting it to int32 would add information that isn't present in the base type.

JonasKlose replied:
@IntelOrca 
Is this ok?

JonasKlose commented:
> Size of windows are just educated guesses usually based on en-gb size. If other languages need more space we do a follow up pr to increase the size. Some windows dynamically resize for the text but there isn't much point wasting performance on that for every window.

Ok. Though it is a bit strange to avoid size calculations on first draw. Even this webpage does it and no one complains about the performance. The calculation would only needed to be done on init or when there are changes in the type of rendered text.

JonasKlose commented:
> Looks good so far, I think it should have percent or progress bar though. Perhaps something like:
> 
> ```
> Combined income:
>     £400 [#####                        ] £1,000
> ```
> 
> using the same progress bars that are drawn on tab 2 of the guest window.
> 
> I also think it should omit the progress section altogether if there is no objective or you have completed / failed it.
> 
> There are some functions that measure string size allowing you update the window size accordingly.

I'll see what I can do about the progress bars. :)
About disabling the progress on a won/failed objective, I've thought about that, too. I'll try to implement it.

JonasKlose commented:
> Looks good so far, I think it should have percent or progress bar though. Perhaps something like:
> 
> ```
> Combined income:
>     £400 [#####                        ] £1,000
> ```
> 
> using the same progress bars that are drawn on tab 2 of the guest window.
> 
> I also think it should omit the progress section altogether if there is no objective or you have completed / failed it.
> 
> There are some functions that measure string size allowing you update the window size accordingly.

The new commit adds a horizontal bar for the guest number and the park rating beneath the text. Check it out and tell me what you think about that. 😸

tupaschoal reviewed (commented):
@JonasKlose can you do us 2 favours, please?

1. Resolve every conversation where you have already changed the code according to the reviewers comments. This makes the PR look cleaner and it tells us that you've tended to them.
2. Add some screenshots showing how it's currently looking for different objectives.

Thanks!

JonasKlose commented:
> @JonasKlose can you do us 2 favours, please?
> 
> 1. Resolve every conversation where you have already changed the code according to the reviewers comments. This makes the PR look cleaner and it tells us that you've tended to them.
> 2. Add some screenshots showing how it's currently looking for different objectives.
> 
> Thanks!

I've resolved all conversations that I think have been resolved. 😃 
Attached at this post are the screenshots of three different scenarios to show you what this branch looks like in game.

![Vertigo Views Test](https://user-images.githubusercontent.com/23458121/104141959-92b01300-53b9-11eb-8acd-de757b3d97aa.png)
![Ghost Town Test](https://user-images.githubusercontent.com/23458121/104141966-980d5d80-53b9-11eb-8c6d-298ddd99d5fd.png)
![Victoria Falls Test](https://user-images.githubusercontent.com/23458121/104141968-9ba0e480-53b9-11eb-8bd5-feef93e8b763.png)


.

James103 commented:
Should the "Your Progress" section stay even after scenario completion (in addition to the victory/failure text if any)? Or should it be removed and replaced with the ending text when the scenario is completed/failed?

JonasKlose commented:
> Should the "Your Progress" section stay even after scenario completion (in addition to the victory/failure text if any)? Or should it be removed and replaced with the ending text when the scenario is completed/failed?

This was suggested earlier and I think you're right. The progress is now only visible when the scenario objective is not completed.

Umdlye commented:
This looks great!

The 'p' in 'progress' should probably be lowercase (like the 'o' in 'objective').

JonasKlose commented:
> This looks great!
> 
> The 'p' in 'progress' should probably be lowercase (like the 'o' in 'objective').

Thanks for the hint! I've updated the PR.

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1446,6 +1446,22 @@ static void window_park_objective_invalidate(rct_window* w)
+    }
+    int16_t barLength = static_cast<int16_t>((221 - 4) * percentage);
+
+    gfx_fill_rect_inset(dpi, x + 1, y + 1, x + 221 - 3, y + 9, w->colours[1], INSET_RECT_F_30);
```
Please use the `gfx_fill_rect_inset` that receives a `ScreenRect` instead.

JonasKlose replied:
Thank you for the hint. I'll update the code accordingly.

JonasKlose replied:
I've switched to using ScreenRects now. :)

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1446,6 +1446,22 @@ static void window_park_objective_invalidate(rct_window* w)
+    gfx_fill_rect_inset(dpi, x + 1, y + 1, x + 221 - 3, y + 9, w->colours[1], INSET_RECT_F_30);
+    if (barLength > 2)
+    {
+        gfx_fill_rect_inset(dpi, x + 3, y + 2, x + barLength, y + 8, colour, 0);
```
Same as above.

JonasKlose replied:
I've switched to using ScreenRects now. :)

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1446,6 +1446,22 @@ static void window_park_objective_invalidate(rct_window* w)
+    {
+        percentage = 1;
+    }
+    int16_t barLength = static_cast<int16_t>((221 - 4) * percentage);
+
+    gfx_fill_rect_inset(dpi, x + 1, y + 1, x + 221 - 3, y + 9, w->colours[1], INSET_RECT_F_30);
```
What is the `221` magic number?

JonasKlose replied:
It just worked out the best for the size of the rectangle to fill the width of the park window.
I don't think the other code that used the function had a proper explanation for its size, too.

JonasKlose replied:
I've looked at the function window_guest_stats_bars_paint in Guest.cpp. There are magic numbers in that code, too. It is used to paint the guest bars and that's the code that I used as example. I used my 221 like the 121 was used there.

Maybe, there's a solution for both functions. But I don't see it. Do you have an idea?

tupaschoal replied:
I think it's alright to keep the magic numbers in this case.

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1446,6 +1446,22 @@ static void window_park_objective_invalidate(rct_window* w)
 }
 
+static void window_park_objective_bars_paint(
+    float percentage, int32_t x, int32_t y, rct_window* w, rct_drawpixelinfo* dpi, int32_t colour)
```
Make it receive a `const ScreenCoordsXY&` object instead of the raw `x` and `y`

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1495,6 +1511,199 @@ static void window_park_objective_paint(rct_window* w, rct_drawpixelinfo* dpi)
+
+                if (gScenarioObjective.Type == OBJECTIVE_GUESTS_BY)
+                {
+                    ft.Add<uint32_t>(static_cast<uint32_t>(600));
+                    parkRatingPercentage = (static_cast<float>(gParkRating) / static_cast<float>(600));
```
What's the `600` magic number? Maybe extract it to a `constexpr` for better documentation

JonasKlose replied:
The 600 and 700 magic number comes from Scenario.cpp. It's used in the functions CheckGuestsBy and CheckGuestsAndRating. I agree a constant would be better. I thought that the whole code of Scenario.cpp could be overhauled, but I guess a step between won't hurt anyone.

JonasKlose replied:
Replaced 600 by the constexpr OBJECTIVE_GUESTS_BY_PARK_RATING.

tupaschoal replied:
Nice!

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1495,6 +1511,199 @@ static void window_park_objective_paint(rct_window* w, rct_drawpixelinfo* dpi)
+                }
+                else
+                {
+                    ft.Add<uint32_t>(static_cast<uint32_t>(700));
+                    parkRatingPercentage = (static_cast<float>(gParkRating) / static_cast<float>(700));
```
What's the `600` magic number? Maybe extract it to a `constexpr` for better documentation

JonasKlose replied:
Replaced 700 by the constexpr OBJECTIVE_GUESTS_AND_RATING_PARK_RATING.

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1495,6 +1511,199 @@ static void window_park_objective_paint(rct_window* w, rct_drawpixelinfo* dpi)
+    // if (gScenarioObjective.Check() == ObjectiveStatus::Undecided)
+    if (gScenarioCompletedCompanyValue == MONEY32_UNDEFINED)
+    {
+        gfx_draw_string_left(dpi, STR_OBJECTIVE_PROGRESS_LABEL, nullptr, COLOUR_BLACK, screenCoords);
```
I think the whole content of this `if` should be extracted to another function, something like `PaintProgressBar`. This is currently >300 lines added to a function that was already big.

JonasKlose replied:
You're right. I'll split the window_park_objective_paint function in four separate functions. It'll be much better to read.

JonasKlose replied:
> I think the whole content of this `if` should be extracted to another function, something like `PaintProgressBar`. This is currently >300 lines added to a function that was already big.

Phew... managed to split the function into four separate functions. I hope this worked out ok.

JonasKlose replied:
Additionally, I've moved the check of the gScenarioCompletedCompanyValue into another function called GetCurrentObjectiveStatus(). This was a magical way of checking the objective status that also worked after the scenario was completet. Now the code is implemented in Scenario.cpp.

JonasKlose commented:
@tupaschoal 
Thank you for all the feedback! 😃

tupaschoal commented:
You're welcome. Let us know when you've tended to the comments and also remember to mark the conversations as resolved when you fix them, thanks!

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1446,31 +1446,43 @@ static void window_park_objective_invalidate(rct_window* w)
 {
-    WindowDrawWidgets(w, dpi);
-    window_park_draw_tab_images(dpi, w);
+    if (percentage > 1)
+    {
+        percentage = 1;
+    }
```
I'm wondering if this wouldn't be more elegant with a `clamp`, to guarantee that it is within boundaries:
```suggestion
    percentage = std::clamp(percentage, 0, 1);
```

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1446,31 +1446,43 @@ static void window_park_objective_invalidate(rct_window* w)
+        percentage = 1;
+    }
+    int16_t barLength = static_cast<int16_t>((221 - 4) * percentage);
+    ScreenRect* rect = new ScreenRect(screenCoords->x + 1, screenCoords->y + 1, screenCoords->x + 221 - 3, screenCoords->y + 9);
```
Don't use pointers, prefer stack allocations:
```suggestion
    auto rect = ScreenRect(screenCoords->x + 1, screenCoords->y + 1, screenCoords->x + 221 - 3, screenCoords->y + 9);
```

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1446,31 +1446,43 @@ static void window_park_objective_invalidate(rct_window* w)
-    gfx_draw_string_left(dpi, STR_OBJECTIVE_LABEL, nullptr, COLOUR_BLACK, screenCoords);
-    screenCoords.y += LIST_ROW_HEIGHT;
+    rect = new ScreenRect(screenCoords->x + 3, screenCoords->y + 2, barLength, screenCoords->y + 8);
+    if (barLength > 2)
+    {
```
Maybe add a comment as to why you're drawing the bar only when length is `> 2`

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1479,40 +1491,263 @@ static void window_park_objective_paint(rct_window* w, rct_drawpixelinfo* dpi)
+                if (gScenarioObjective.Type == OBJECTIVE_GUESTS_BY)
+                {
+                    ft->Add<uint32_t>(static_cast<uint32_t>(OBJECTIVE_GUESTS_BY_PARK_RATING));
+                    parkRatingPercentage = (static_cast<float>(gParkRating) / static_cast<float>(600));
```
```suggestion
                    parkRatingPercentage = (static_cast<float>(gParkRating) / static_cast<float>(OBJECTIVE_GUESTS_BY_PARK_RATING));
```

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1479,40 +1491,263 @@ static void window_park_objective_paint(rct_window* w, rct_drawpixelinfo* dpi)
+                else
+                {
+                    ft->Add<uint32_t>(static_cast<uint32_t>(OBJECTIVE_GUESTS_AND_RATING_PARK_RATING));
+                    parkRatingPercentage = (static_cast<float>(gParkRating) / static_cast<float>(700));
```
```suggestion
                    parkRatingPercentage = (static_cast<float>(gParkRating) / static_cast<float>(OBJECTIVE_GUESTS_AND_RATING_PARK_RATING));
```

tupaschoal commented on src/openrct2-ui/windows/Park.cpp:
```diff
@@ -1479,40 +1491,263 @@ static void window_park_objective_paint(rct_window* w, rct_drawpixelinfo* dpi)
+            {
+                float guestsInParkPercentage
+                    = (static_cast<float>(gNumGuestsInPark) / static_cast<float>(gScenarioObjective.NumGuests));
+                float parkRatingPercentage;
```
Move this line to just before the `if (gScenarioObjective.Type == OBJECTIVE_GUESTS_BY)`

You can even initialize it to `static_cast<float>(gParkRating)` and then just divide it inside the `if` or `else`

tupaschoal reviewed (commented):
I've just done a partial review, but in general, try to avoid dynamic allocations like the `new ScreenRect` or `new ScreenCoordsXY`.

Also, it would be great to have most of the coordinate parameters as well as the formatter be const references, like it is in most of the code.

Gymnasiast commented:
Closing this due to inactivity. If you ever decide to pick it up again, you can open a new one. Do make sure you address the outstanding line notes if you do so.
